[Step2] 문자열 계산기

misudev opened the pull request:
안녕하세요 리뷰어님!
2단계 리뷰 부탁드립니다! 🙌

misudev committed:
test: 사칙연산 테스트 작성

misudev committed:
test: 수식 계산, 예외 케이스 테스트 작성

misudev committed:
feat: 문자열 계산기 구현

misudev committed:
refactor: step1 패키지 이동

kyucumber commented on src/main/kotlin/step2/Operator.kt:
```diff
@@ -0,0 +1,26 @@
+
+import java.util.function.BiFunction
+
+enum class Operator(val value: String) : BiFunction<Int, Int, Int> {
```
BiFunction 인터페이스를 잘 활용해주셨네요. 인터페이스 대신 함수 타입을 사용해보시는건 어떨까요?

함수 타입을 사용하면 조금 더 많은 자유를 누릴 수 있습니다. 자세한 내용은 아래 문서 참고 부탁드립니다.

[이펙티브 코틀린 아이템 38. 연산 또는 액션을 전달할 때는 인터페이스 대신 함수 타입을 사용하라](https://kyucumber.github.io/posts/book/effective-kotlin/effective-kotlin-item-38)

kyucumber commented on src/main/kotlin/step2/Operator.kt:
```diff
@@ -0,0 +1,26 @@
+import java.util.function.BiFunction
+
+enum class Operator(val value: String) : BiFunction<Int, Int, Int> {
+    PLUS("+") {
+        override fun apply(t: Int, u: Int): Int = t + u
+    },
+
+    MINUS("-") {
+        override fun apply(t: Int, u: Int): Int = t - u
+    },
+
+    MULTIPLY("*") {
+        override fun apply(t: Int, u: Int): Int = t * u
+    },
+
+    DIVIDE("/") {
+        override fun apply(t: Int, u: Int): Int = t / u
+    };
```
if, else 구문을 잘 제거해주셨네요 👍

kyucumber commented on src/main/kotlin/step2/Operator.kt:
```diff
@@ -0,0 +1,26 @@
+
+    companion object {
+        private val operators = Operator.values().associateBy(Operator::value)
+        fun from(value: String) = operators[value]
```
Nullable 타입이 반환될 것 같네요 사용처의 코드를 보면 아래와 같습니다.

```kotlin
            val operator = Operator.from(inputArray[i - 1]) ?: throw IllegalArgumentException()
```

예외처리를 매번 사용하는 곳에서 하기 보다는 Non Null 타입을 반환하고 from 메소드에서 예외처리를 하는게 어떨까요?

kyucumber commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,19 @@
+        var result = inputArray[0].toInt()
+
+        for (i in 2 until inputArray.size step 2) {
+            val operator = Operator.from(inputArray[i - 1]) ?: throw IllegalArgumentException()
```
operator는 코틀린의 예약어입니다. 가급적 다른 변수명을 사용해주시면 어떨까요?

kyucumber commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,19 @@
+
+class StringCalculator {
+    fun calculate(input: String?): Int {
+        if (input.isNullOrEmpty()) throw IllegalArgumentException()
```
코틀린에는 IllegalArgumentException을 처리하기 위한 require라는 stdlib 메소드가 존재합니다.

require 메소드를 활용해보세요.

관련 내용은 [이펙티브 코틀린 아이템 05. 예외를 활용해 코드에 제한을 걸어라](https://kyucumber.github.io/posts/book/effective-kotlin/effective-kotlin-item-05)를 참고해보세요.

kyucumber commented on src/test/kotlin/step2/StringCalculatorTest.kt:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    @Test
+    fun `입력값이 null인 경우`() {
```
Junit의 NullSource와 ParameterizedTest를 활용해보셔도 좋을 것 같아요.

kyucumber commented on src/test/kotlin/step2/StringCalculatorTest.kt:
```diff
@@ -0,0 +1,67 @@
+import org.junit.jupiter.api.Test
+
+class StringCalculatorTest {
+    @Test
+    fun `덧셈`() {
+        val input = "1 + 2"
+        val stringCalculator = StringCalculator()
+
+        assertThat(stringCalculator.calculate(input)).isEqualTo(3)
+    }
+
+    @Test
+    fun `뺄셈`() {
+        val input = "10 - 2"
+        val stringCalculator = StringCalculator()
+
+        assertThat(stringCalculator.calculate(input)).isEqualTo(8)
+    }
+
+    @Test
+    fun `곱셈`() {
+        val input = "5 * 2"
+        val stringCalculator = StringCalculator()
+
+        assertThat(stringCalculator.calculate(input)).isEqualTo(10)
+    }
+
+    @Test
+    fun `나눗셈`() {
+        val input = "10 / 2"
+        val stringCalculator = StringCalculator()
+
+        assertThat(stringCalculator.calculate(input)).isEqualTo(5)
+    }
```
같은 연산에 input만 다른 것 같은데 ParameterizedTest를 활용해 하나로 합쳐볼 수 있을 것 같아요.

kyucumber reviewed (changes requested):
안녕하세요 몇가지 코멘트 남겨두었으니 확인 부탁드립니다.

misudev committed:
fix: 코드리뷰 반영

- 인터페이스 대신 함수 사용하기
- from() 에서 non null 처리
- 코틀린의 예약어 사용하지 않기
- require 사용하기
- ParameterizedTest 사용, 중복코드 줄이기

misudev commented:
> 안녕하세요 몇가지 코멘트 남겨두었으니 확인 부탁드립니다.

꼼꼼한 리뷰 감사드립니다! 리뷰 주신부분 수정하였습니다 :)

kyucumber reviewed (approved):
수고 많으셨습니다. 
다음 단계 진행을 위해 머지할게요.
